### ISSUE TITLE
Improve code quality

### ISSUE THREAD

**Issue by genodeftest** [exaile/exaile#371]
Currently, there are quite many warnings and errors when Exaile's python code is analyzed by `pep8` or `pylint`. I would like to improve the code quality.

### Steps to Reproduce (for bugs)
Run code analysis on Exaile code – or just open the files and read code.

### Expected Behavior
Clean, well-formatted code. According to [doc/dev/code_guidelines.rst](https://github.com/exaile/exaile/blob/master/doc/dev/code_guidelines.rst) the code should comply to pep8

### Current Behavior
Many warnings when running code analysis tools such as pep8 or pylint.

### Possible Solution
[ ] `pep8` and `autopep8`:
- [ ] run `autopep8` over the whole tree
- [ ] run `pep8` and manually fix remaining warnings
- [ ] add a `pep8`-based code checker to continuous integration tests
[ ] `pyflakes`, `pychecker`
[ ] `pylint` – repeat steps above (?)

### Environment
* Exaile version: current git master

### Additional info:
https://flake8.readthedocs.io/en/latest/
https://pycodestyle.readthedocs.io/en/latest/intro.html#error-codes

### Discussion:
In [a comment](https://github.com/exaile/exaile/pull/310#issuecomment-267244829), @virtuald voted against enforcing the 80-characters rule too strong. Is there a consensus for that? If yes, we should change the docs.

**Comment by virtuald** [comment]
I think if you're going to take the time to clean it up, then we should add some tests to travis to enforce it on new code too.. otherwise it'll just get messy again?

I'm still very against an 80-character limit. I don't mind sensible code style that helps readability, but sometimes it just gets in the way.

**Comment by genodeftest** [comment]
> I think if you're going to take the time to clean it up, then we should add some tests to travis to enforce it on new code too.. otherwise it'll just get messy again?

That's what I wrote above. I don't know how to do this (yet), but there will be a way. I'll do that after the manual cleanup.

**Comment by genodeftest** [comment]
> I'm still very against an 80-character limit. I don't mind sensible code style that helps readability, but sometimes it just gets in the way.

So, is a 100-character limit ok for you? or 120?

**Comment by genodeftest** [comment]
With pycodestyle (formerly called pep8), I'd like to ignore these "errors" for now:
* E241 – multiple spaces after ‘,’
* E265 – block comment should start with ‘# ‘
* E266 – too many leading ‘#’ for block comment
* E402 – module level import not at top of file
* E501 – line too long (82 > 79 characters)

See https://pycodestyle.readthedocs.io/en/latest/intro.html#error-codes for a full list of error codes and their meaning. I'll commit a modified `.travis.yml` after #372 is merged.

**Comment by javiercantero** [comment]
One of the problems with these massive patches that touch almost every file in the source tree is that they generate artificial conflicts when merging any branch from the past. So my unrequested advice is to merge every pending PR/WIP branch before applying this to avoid further work.

**Comment by genodeftest** [comment]
> One of the problems with these massive patches that touch almost every file in the source tree is that they generate artificial conflicts when merging any branch from the past. So my unrequested advice is to merge every pending PR/WIP branch before applying this to avoid further work.
True. Since the pull requests are easily redone, there is no major problem with that. On the other hand, #245 and #256 already have merge conflicts and haven't been rebased nor merged for >1 year.

**Comment by javiercantero** [comment]
These are the different errors and warnings (sorted) that pycodestyle 2.2.0 is reporting over `xl/` and `xlgui/`:

> E101 indentation contains mixed spaces and tabs
> E111 indentation is not a multiple of four
> E114 indentation is not a multiple of four (comment)
> E115 expected an indented block (comment)
> E116 unexpected indentation (comment)
> E122 continuation line missing indentation or outdented
> E124 closing bracket does not match visual indentation
> E125 continuation line with same indent as next logical line
> E127 continuation line over-indented for visual indent
> E128 continuation line under-indented for visual indent
> E129 visually indented line with same indent as next logical line
> E131 continuation line unaligned for hanging indent
> E201 whitespace after '('
> E202 whitespace before ')'
> E203 whitespace before ':'
> E221 multiple spaces before operator
> E225 missing whitespace around operator
> E227 missing whitespace around bitwise or shift operator
> E228 missing whitespace around modulo operator
> E231 missing whitespace after ','
> E251 unexpected spaces around keyword / parameter equals
> E261 at least two spaces before inline comment
> E262 inline comment should start with '# '
> E265 block comment should start with '# '
> E266 too many leading '#' for block comment
> E271 multiple spaces after keyword
> E272 multiple spaces before keyword
> E301 expected 1 blank line, found 0
> E302 expected 2 blank lines, found 1
> E303 too many blank lines (3)
> E305 expected 2 blank lines after class or function definition, found 3
> E306 expected 1 blank line before a nested definition, found 0
> E401 multiple imports on one line
> E402 module level import not at top of file
> E501 line too long (102 > 100 characters)
> E502 the backslash is redundant between brackets
> E701 multiple statements on one line (colon)
> E703 statement ends with a semicolon
> E711 comparison to None should be 'if cond is None:'
> E712 comparison to True should be 'if cond is True:' or 'if cond:'
> E713 test for membership should be 'not in'
> E731 do not assign a lambda expression, use a def
> E741 ambiguous variable name 'l'
> 
> W191 indentation contains tabs
> W291 trailing whitespace
> W293 blank line contains whitespace
> W391 blank line at end of file

And applied to the entire source tree there are also:

> E211 whitespace before '('
> E222 multiple spaces after operator
> E721 do not compare types, use 'isinstance()'

Can I suggest to pick the low hanging fruit first (things like E703, E711, E712, E713), and start by changing only the core files under `xl/` and `xlgui/`? That way:
 1. the changes are not massive
 2. later it will be easy to follow the history of each file

**Comment by javiercantero** [comment]
I've done #373 as a proof-of-concept of the above idea. I can add more checks if you all are ok with it.

**Comment by virtuald** [comment]
@javiercantero those seem like a sensible subset to enforce

**Comment by genodeftest** [comment]
Out of the list of default enabled fixers of `autopep8`, these are not applied in exaile source yet:

* W291 trailing whitespace (#374)
* E712 comparison to True should be 'if cond is True:' or 'if cond:' (#374, see also #373)

* E26: comment-related fixers (See b89c53d)
  * Question: Do you want this?
* E2: remaining fixers about whitespace in lines with text (See b027c5f)
  * Question: Do you want this?
* W293 blank line contains whitespace
  * Removes all whitespace from empty lines
  * Question: Do you want this?
* E501 line too long
  * Limits line length
  * Question: Do you want this?
  * Question: How long should lines be at maximum? (Default: 80 characters) How about 100 or 120 chars?

**Comment by javiercantero** [comment]
After #378 these pycodestyle warnings remain:

```
1       E125 continuation line with same indent as next logical line
110     E203 whitespace before ':'
22      E221 multiple spaces before operator
1       E262 inline comment should start with '# '
102     E265 block comment should start with '# '
44      E266 too many leading '#' for block comment
114     E305 expected 2 blank lines after class or function definition, found 1
103     E402 module level import not at top of file
1741    E501 line too long (81 > 79 characters)
10      E731 do not assign a lambda expression, use a def
8       E741 ambiguous variable name 'l'
```

I haven't fixed E221 because it's a very common formatting idiom to vertically align the `=` operator in blocks of assignments. The same happens with E221 and the `:` operator used to define several dictionaries. I don't think that changing any of those would improve readability.

I have my doubts about how to treat E741 and E731.
